File: zig/tests/golden.zig
//! Cross-language golden scenario runner: executes the scenario described
//! by golden/scenario.txt (path injected via build options) and asserts the
//! emitted trace matches golden/expected.txt exactly. The Go and Rust ports
//! ship structurally identical runners; see golden/README.md.

const std = @import("std");
const cordis = @import("cordis");

// The shared scenario and its expected trace, embedded at build time from
// ../golden/ so the runner needs no runtime file access.
const golden_data = @import("golden_data");

const Context = cordis.Context;
const Fiber = cordis.Fiber;

const ChildSpec = struct {
    name: []const u8,
    deps: []const []const u8 = &.{},
    config: i32 = 0,
};

const Runner = struct {
    gpa: std.mem.Allocator,
    trace: std.ArrayListUnmanaged([]const u8) = .empty,
    fibers: std.StringHashMapUnmanaged(Fiber) = .empty,
    plugins: std.StringHashMapUnmanaged(*const cordis.Plugin) = .empty,
    children: std.StringHashMapUnmanaged([]const ChildSpec) = .empty,

    fn log(self: *Runner, comptime format: []const u8, args: anytype) void {
        const line = std.fmt.allocPrint(self.gpa, format, args) catch @panic("cordis: out of memory");
        self.trace.append(self.gpa, line) catch @panic("cordis: out of memory");
    }

    fn applyPlugin(self: *Runner, c: *Context, config: i32, body: *const PluginBody) cordis.Error!void {
        const name = body.name;
        self.log("apply {s} config={d}", .{ name, config });
        var i: usize = 1;
        const count: usize = if (body.lifo) 3 else 1;
        while (i <= count) : (i += 1) {
            const label = if (body.lifo)
                std.fmt.allocPrint(self.gpa, "{s}#{d}", .{ name, i }) catch @panic("cordis: out of memory")
            else
                name;
            const entry = self.gpa.create(CleanupEntry) catch @panic("cordis: out of memory");
            entry.* = .{ .runner = self, .label = label };
            try c.attach(entry, CleanupEntry.run);
        }
        for (body.children) |spec| {
            startLogical(self, c, spec.name, .{ .deps = spec.deps, .config = spec.config }) catch @panic("golden: child start failed");
        }
    }

    fn applyProvider(c: *Context, service: []const u8) cordis.Error!void {
        const stored = c.core.a().create(i32) catch @panic("cordis: out of memory");
        stored.* = 1;
        _ = try c.provideNamed(service, cordis.value(stored));
    }
};

const CleanupEntry = struct {
    runner: *Runner,
    label: []const u8,

    fn run(entry: *CleanupEntry) void {
        entry.runner.log("cleanup {s}", .{entry.label});
    }
};

// Runtime plugin plumbing: each logical plugin owns a PluginBody stored in
// the tree arena and referenced through Plugin.data, so applies always see
// their own name even when dependency reactivation defers them.
const PluginBody = struct {
    runner: *Runner,
    name: []const u8,
    lifo: bool = false,
    children: []const ChildSpec = &.{},
};

const ProviderBody = struct {
    service: []const u8,
};

fn genericApply(plugin: *const cordis.Plugin, c: *Context, raw: ?cordis.Value) cordis.Error!void {
    const body: *const PluginBody = @ptrCast(@alignCast(plugin.data.?));
    const config: i32 = if (raw) |v| @as(*const i32, @ptrCast(@alignCast(v))).* else 0;
    return body.runner.applyPlugin(c, config, body);
}

fn providerApply(plugin: *const cordis.Plugin, c: *Context, _: ?cordis.Value) cordis.Error!void {
    const body: *const ProviderBody = @ptrCast(@alignCast(plugin.data.?));
    return Runner.applyProvider(c, body.service);
}

const Params = struct {
    deps: []const []const u8 = &.{},
    realm: []const u8 = "",
    config: i32 = 0,
    lifo: bool = false,
};

fn parseParams(gpa: std.mem.Allocator, tokens: []const []const u8) Params {
    var p = Params{};
    for (tokens) |tok| {
        if (std.mem.startsWith(u8, tok, "inject=")) {
            const rest = tok["inject=".len..];
            var deps: std.ArrayListUnmanaged([]const u8) = .empty;
            var it = std.mem.splitScalar(u8, rest, ',');
            while (it.next()) |dep| {
                if (dep.len > 0) deps.append(gpa, dep) catch @panic("cordis: out of memory");
            }
            p.deps = deps.toOwnedSlice(gpa) catch @panic("cordis: out of memory");
        } else if (std.mem.startsWith(u8, tok, "realm=")) {
            p.realm = tok["realm=".len..];
        } else if (std.mem.startsWith(u8, tok, "config=")) {
            p.config = std.fmt.parseInt(i32, tok["config=".len..], 10) catch @panic("bad config");
        } else if (std.mem.eql(u8, tok, "lifo")) {
            p.lifo = true;
        }
    }
    return p;
}

fn stateName(state: cordis.FiberState) []const u8 {
    return switch (state) {
        .pending => "PENDING",
        .loading => "LOADING",
        .active => "ACTIVE",
        .failed => "FAILED",
        .disposed => "DISPOSED",
        .unloading => "UNLOADING",
    };
}

fn getOrCreatePlugin(r: *Runner, ctx: *Context, name: []const u8, params: Params) *const cordis.Plugin {
    if (r.plugins.get(name)) |p| return p;
    const body = ctx.core.a().create(PluginBody) catch @panic("cordis: out of memory");
    body.* = .{
        .runner = r,
        .name = name,
        .lifo = params.lifo,
        .children = r.children.get(name) orelse &.{},
    };
    const p = ctx.core.a().create(cordis.Plugin) catch @panic("cordis: out of memory");
    p.* = .{ .name = name, .inject = params.deps, .apply = genericApply, .data = body };
    r.plugins.put(r.gpa, name, p) catch @panic("cordis: out of memory");
    return p;
}

fn startLogical(r: *Runner, ctx: *Context, name: []const u8, params: Params) !void {
    const p = getOrCreatePlugin(r, ctx, name, params);
    const config = ctx.core.a().create(i32) catch @panic("cordis: out of memory");
    config.* = params.config;
    const fiber = try p.start(ctx, cordis.value(config));
    try r.fibers.put(r.gpa, name, fiber);
}

fn startProvider(r: *Runner, ctx: *Context, service: []const u8) !void {
    const body = ctx.core.a().create(ProviderBody) catch @panic("cordis: out of memory");
    body.* = .{ .service = service };
    const name = std.fmt.allocPrint(ctx.core.a(), "provider:{s}", .{service}) catch @panic("cordis: out of memory");
    const p = ctx.core.a().create(cordis.Plugin) catch @panic("cordis: out of memory");
    p.* = .{ .name = name, .apply = providerApply, .data = body };
    const fiber = try p.start(ctx, null);
    const key = std.fmt.allocPrint(r.gpa, "provider:{s}", .{service}) catch @panic("cordis: out of memory");
    try r.fibers.put(r.gpa, key, fiber);
}

fn readScenarioLines(gpa: std.mem.Allocator, text: []const u8) ![][]const u8 {
    var lines: std.ArrayListUnmanaged([]const u8) = .empty;
    var it = std.mem.splitScalar(u8, text, '\n');
    while (it.next()) |raw| {
        const line = std.mem.trim(u8, raw, " \t\r");
        if (line.len == 0 or line[0] == '#') continue;
        try lines.append(gpa, line);
    }
    return lines.toOwnedSlice(gpa);
}

test "golden scenario events" {
    const gpa = std.testing.allocator;
    var arena = std.heap.ArenaAllocator.init(gpa);
    defer arena.deinit();
    const alloc = arena.allocator();

    const scenario = try readScenarioLines(alloc, golden_data.scenario_events);
    const expected = try readScenarioLines(alloc, golden_data.expected_events);

    const ctx = try Context.init(gpa);
    defer ctx.deinit();

    var r = Runner{ .gpa = alloc };

    const Sink = struct {
        runner: *Runner,
        prefix: []const u8,

        fn fire(self: *@This(), args: []const cordis.Value) ?cordis.Value {
            const payload: i32 = if (args.len > 0)
                @as(*const i32, @ptrCast(@alignCast(args[0]))).*
            else
                0;
            self.runner.log("{s}{d}", .{ self.prefix, payload });
            return null;
        }
    };

    for (scenario) |line| {
        var tokens = std.mem.tokenizeAny(u8, line, " \t");
        const op = tokens.next().?;
        const event = tokens.next().?;
        var payload: i32 = 0;
        var realm: []const u8 = "";
        while (tokens.next()) |tok| {
            if (std.mem.startsWith(u8, tok, "payload=")) {
                payload = std.fmt.parseInt(i32, tok["payload=".len..], 10) catch @panic("bad payload");
            } else if (std.mem.startsWith(u8, tok, "realm=")) {
                realm = tok["realm=".len..];
            }
        }

        if (std.mem.eql(u8, op, "on") or std.mem.eql(u8, op, "on-isolated")) {
            var scope: *Context = ctx;
            var prefix: []const u8 = undefined;
            if (realm.len > 0) {
                scope = ctx.isolateShared(event, realm);
                prefix = std.fmt.allocPrint(alloc, "fired {s} realm={s} payload=", .{ event, realm }) catch @panic("cordis: out of memory");
            } else {
                prefix = std.fmt.allocPrint(alloc, "fired {s} root payload=", .{event}) catch @panic("cordis: out of memory");
            }
            const sink = alloc.create(Sink) catch @panic("cordis: out of memory");
            sink.* = .{ .runner = &r, .prefix = prefix };
            _ = try scope.onNamed(event, cordis.Listener.bind(Sink, sink, Sink.fire));
        } else if (std.mem.eql(u8, op, "on-global")) {
            const prefix = std.fmt.allocPrint(alloc, "fired {s} global payload=", .{event}) catch @panic("cordis: out of memory");
            const sink = alloc.create(Sink) catch @panic("cordis: out of memory");
            sink.* = .{ .runner = &r, .prefix = prefix };
            _ = try ctx.onGlobal(event, cordis.Listener.bind(Sink, sink, Sink.fire));
        } else if (std.mem.eql(u8, op, "emit")) {
            const stored = try ctx.core.a().create(i32);
            stored.* = payload;
            ctx.emitNamed(event, &.{cordis.value(stored)});
        } else if (std.mem.eql(u8, op, "emit-filtered")) {
            const scope = ctx.isolateShared(event, realm);
            const emitter = scope.withFilter(scope.realmFilter(scope, event));
            const stored = try ctx.core.a().create(i32);
            stored.* = payload;
            emitter.emitNamed(event, &.{cordis.value(stored)});
        } else {
            std.debug.print("unknown op {s}\n", .{op});
            return error.GoldenUnknownOp;
        }
    }

    if (r.trace.items.len != expected.len) {
        std.debug.print("trace length {d} != expected {d}\ntrace:\n", .{ r.trace.items.len, expected.len });
        for (r.trace.items) |entry| std.debug.print("{s}\n", .{entry});
        return error.GoldenLengthMismatch;
    }
    for (r.trace.items, expected, 0..) |got, want, i| {
        if (!std.mem.eql(u8, got, want)) {
            std.debug.print("trace divergence at line {d}: expected '{s}', got '{s}'\n", .{ i + 1, want, got });
            return error.GoldenDivergence;
        }
    }
}

test "golden scenario" {
    const gpa = std.testing.allocator;
    var arena = std.heap.ArenaAllocator.init(gpa);
    defer arena.deinit();
    const alloc = arena.allocator();

    const scenario = try readScenarioLines(alloc, golden_data.scenario);
    const expected = try readScenarioLines(alloc, golden_data.expected);

    const ctx = try Context.init(gpa);
    defer ctx.deinit();

    var r = Runner{ .gpa = alloc };

    for (scenario) |line| {
        var tokens = std.mem.tokenizeAny(u8, line, " \t");
        const op = tokens.next().?;
        var args: std.ArrayListUnmanaged([]const u8) = .empty;
        while (tokens.next()) |tok| try args.append(alloc, tok);

        if (std.mem.eql(u8, op, "provide") or std.mem.eql(u8, op, "provide-in-realm")) {
            const params = parseParams(alloc, args.items);
            var scope: *Context = ctx;
            if (params.realm.len > 0) scope = ctx.isolateShared(args.items[0], params.realm);
            try startProvider(&r, scope, args.items[0]);
            r.log("provided {s}", .{args.items[0]});
        } else if (std.mem.eql(u8, op, "withdraw") or std.mem.eql(u8, op, "withdraw-in-realm")) {
            const key = try std.fmt.allocPrint(alloc, "provider:{s}", .{args.items[0]});
            const fiber = r.fibers.get(key).?;
            fiber.dispose();
            r.log("withdrawn {s}", .{args.items[0]});
        } else if (std.mem.eql(u8, op, "start") or std.mem.eql(u8, op, "start-isolated")) {
            const params = parseParams(alloc, args.items);
            var scope: *Context = ctx;
            if (std.mem.eql(u8, op, "start-isolated")) {
                for (params.deps) |dep| scope = scope.isolateShared(dep, params.realm);
            }
            try startLogical(&r, scope, args.items[0], params);
        } else if (std.mem.eql(u8, op, "update")) {
            const fiber = r.fibers.get(args.items[0]).?;
            const config = try std.fmt.parseInt(i32, args.items[1], 10);
            const stored = try ctx.core.a().create(i32);
            stored.* = config;
            try fiber.update(cordis.value(stored));
        } else if (std.mem.eql(u8, op, "restart")) {
            try r.fibers.get(args.items[0]).?.restart();
        } else if (std.mem.eql(u8, op, "dispose")) {
            r.fibers.get(args.items[0]).?.dispose();
            r.log("disposed {s}", .{args.items[0]});
        } else if (std.mem.eql(u8, op, "restart-root")) {
            ctx.fiberHandle().dispose();
            r.log("root-restarted", .{});
        } else if (std.mem.eql(u8, op, "spawn")) {
            var parent: []const u8 = "";
            for (args.items) |tok| {
                if (std.mem.startsWith(u8, tok, "parent=")) parent = tok["parent=".len..];
            }
            const params = parseParams(alloc, args.items[1..]);
            const spec = alloc.create(ChildSpec) catch @panic("cordis: out of memory");
            spec.* = .{ .name = args.items[0], .deps = params.deps, .config = params.config };
            const existing = r.children.get(parent) orelse &.{};
            var list: std.ArrayListUnmanaged(ChildSpec) = .empty;
            list.appendSlice(alloc, existing) catch @panic("cordis: out of memory");
            list.append(alloc, spec.*) catch @panic("cordis: out of memory");
            r.children.put(alloc, parent, list.toOwnedSlice(alloc) catch @panic("cordis: out of memory")) catch @panic("cordis: out of memory");
        } else if (std.mem.eql(u8, op, "delete")) {
            const p = r.plugins.get(args.items[0]).?;
            ctx.registry().delete(p);
            r.log("deleted {s}", .{args.items[0]});
        } else if (std.mem.eql(u8, op, "expect-registry-size")) {
            const want = try std.fmt.parseInt(usize, args.items[0], 10);
            const got = ctx.registry().size();
            if (got != want) {
                std.debug.print("expected registry size {d}, got {d}\n", .{ want, got });
                return error.GoldenRegistrySizeMismatch;
            }
            r.log("registry-size {d}", .{want});
        } else if (std.mem.eql(u8, op, "expect-state")) {
            const fiber = r.fibers.get(args.items[0]).?;
            const got = stateName(fiber.state());
            if (!std.mem.eql(u8, got, args.items[1])) {
                std.debug.print("expected {s} {s}, got {s}\ntrace:\n", .{ args.items[0], args.items[1], got });
                for (r.trace.items) |entry| std.debug.print("{s}\n", .{entry});
                return error.GoldenStateMismatch;
            }
            r.log("state {s} {s}", .{ args.items[0], args.items[1] });
        } else {
            std.debug.print("unknown op {s}\n", .{op});
            return error.GoldenUnknownOp;
        }
    }

    if (r.trace.items.len != expected.len) {
        std.debug.print("trace length {d} != expected {d}\ntrace:\n", .{ r.trace.items.len, expected.len });
        for (r.trace.items) |entry| std.debug.print("{s}\n", .{entry});
        return error.GoldenLengthMismatch;
    }
    for (r.trace.items, expected, 0..) |got, want, i| {
        if (!std.mem.eql(u8, got, want)) {
            std.debug.print("trace divergence at line {d}: expected '{s}', got '{s}'\n", .{ i + 1, want, got });
            return error.GoldenDivergence;
        }
    }
}

test "golden dsl parseParams" {
    const alloc = std.testing.allocator;
    var arena = std.heap.ArenaAllocator.init(alloc);
    defer arena.deinit();
    const a = arena.allocator();

    const tokens = [_][]const u8{ "inject=a,,b", "realm=tenant", "config=7", "lifo", "ignored" };
    const p = parseParams(a, &tokens);
    try std.testing.expectEqual(@as(usize, 2), p.deps.len);
    try std.testing.expectEqualStrings("a", p.deps[0]);
    try std.testing.expectEqualStrings("b", p.deps[1]);
    try std.testing.expectEqualStrings("tenant", p.realm);
    try std.testing.expectEqual(@as(i32, 7), p.config);
    try std.testing.expect(p.lifo);

    const empty = parseParams(a, &.{});
    try std.testing.expectEqual(@as(usize, 0), empty.deps.len);
    try std.testing.expectEqual(@as(i32, 0), empty.config);
    try std.testing.expect(!empty.lifo);
}

test "golden dsl scenario line parsing" {
    const alloc = std.testing.allocator;
    var arena = std.heap.ArenaAllocator.init(alloc);
    defer arena.deinit();
    const a = arena.allocator();

    const lines = try readScenarioLines(a, "# comment\n\nstart worker inject=a config=1\n");
    try std.testing.expectEqual(@as(usize, 1), lines.len);
    try std.testing.expectEqualStrings("start worker inject=a config=1", lines[0]);

    var it = std.mem.tokenizeAny(u8, lines[0], " \t");
    try std.testing.expectEqualStrings("start", it.next().?);
    try std.testing.expectEqualStrings("worker", it.next().?);
}

File: rust/tests/golden.rs
// Tests legitimately assert via panic; the strict production
// lints (unwrap/expect/indexing/arithmetic) are relaxed here.
#![allow(clippy::unwrap_used, clippy::expect_used, clippy::indexing_slicing, clippy::arithmetic_side_effects, clippy::panic)]

//! Cross-language golden scenario runners: executes ../golden/scenario*.txt
//! and asserts the emitted traces match the ../golden/expected*.txt files
//! exactly. The Go and Zig ports ship structurally identical runners; see
//! golden/README.md.

use cordis::sync::RefCell;
use std::collections::HashMap;
use std::fs;
use std::path::PathBuf;
use cordis::sync::{BorrowExt as _, Rc};

use cordis::{plugin, start_fn, value, Context, EventOptions, Fiber, FiberState, FnPlugin, Listener, Value};

type Trace = Rc<RefCell<Vec<String>>>;
type FiberMap = Rc<RefCell<HashMap<String, Fiber>>>;
type PluginMap = Rc<RefCell<HashMap<String, Rc<FnPlugin<i32>>>>>;

/// Deferred child plugins started inside a parent plugin's body.
#[derive(Clone)]
struct Spawn {
    name: String,
    deps: Vec<String>,
    config: i32,
}

struct Runner {
    ctx: Context,
    trace: Trace,
    fibers: FiberMap,
    plugins: PluginMap,
    children: HashMap<String, Vec<Spawn>>,
}

fn state_name(state: FiberState) -> &'static str {
    match state {
        FiberState::Pending => "PENDING",
        FiberState::Loading => "LOADING",
        FiberState::Active => "ACTIVE",
        FiberState::Failed => "FAILED",
        FiberState::Disposed => "DISPOSED",
        FiberState::Unloading => "UNLOADING",
    }
}

#[derive(Default)]
struct Params {
    deps: Vec<String>,
    realm: String,
    config: i32,
    lifo: bool,
}

fn parse_params(tokens: &[String]) -> Params {
    let mut p = Params::default();
    for tok in tokens {
        if let Some(rest) = tok.strip_prefix("inject=") {
            p.deps = rest.split(',').filter(|d| !d.is_empty()).map(String::from).collect();
        } else if let Some(rest) = tok.strip_prefix("realm=") {
            p.realm = rest.to_string();
        } else if let Some(rest) = tok.strip_prefix("config=") {
            p.config = rest.parse().expect("config int");
        } else if let Some(rest) = tok.strip_prefix("parent=") {
            p.realm = rest.to_string();
        } else if tok == "lifo" {
            p.lifo = true;
        }
    }
    p
}

fn golden_dir() -> PathBuf {
    PathBuf::from(env!("CARGO_MANIFEST_DIR")).join("../golden")
}

fn read_lines(name: &str) -> Vec<String> {
    let text = fs::read_to_string(golden_dir().join(name)).expect(name);
    text.lines()
        .map(str::trim)
        .filter(|l| !l.is_empty() && !l.starts_with('#'))
        .map(String::from)
        .collect()
}

/// Split a scenario line into op and args.
fn tokenize(line: &str) -> (String, Vec<String>) {
    let mut tokens = line.split_whitespace().map(String::from);
    let op = tokens.next().expect("op");
    (op, tokens.collect())
}

fn compare_trace(trace: &[String], expected: &[String]) {
    assert_eq!(
        trace.len(),
        expected.len(),
        "trace length mismatch\ntrace:\n{}\nexpected:\n{}",
        trace.join("\n"),
        expected.join("\n")
    );
    for (i, (got, want)) in trace.iter().zip(expected.iter()).enumerate() {
        assert_eq!(got, want, "trace divergence at line {}", i + 1);
    }
}

/// Build an uncached plugin whose body traces apply/cleanup and starts any
/// spawned children on its own context.
fn make_plugin(
    name: &str,
    lifo: bool,
    spawns: Vec<Spawn>,
    trace: &Trace,
    fibers: &FiberMap,
) -> FnPlugin<i32> {
    let trace = Rc::clone(trace);
    let fibers = Rc::clone(fibers);
    let name = name.to_string();
    let plugin_name = name.clone();
    plugin(plugin_name.as_str(), move |ctx: &Context, config: &i32| {
        trace.borrow_mut().push(format!("apply {name} config={config}"));
        let labels: Vec<String> = if lifo {
            (1..=3).map(|i| format!("{name}#{i}")).collect()
        } else {
            vec![name.clone()]
        };
        for label in labels {
            let trace = Rc::clone(&trace);
            ctx.attach(move || {
                trace.borrow_mut().push(format!("cleanup {label}"));
            })?;
        }
        for spawn in spawns.iter() {
            let child = make_plugin(&spawn.name, false, Vec::new(), &trace, &fibers);
            let deps: Vec<&str> = spawn.deps.iter().map(String::as_str).collect();
            let child = if deps.is_empty() { child } else { child.inject(&deps) };
            let fiber = start_fn(ctx, &child, spawn.config)?;
            fibers.borrow_mut().insert(spawn.name.clone(), fiber);
        }
        Ok(())
    })
}

impl Runner {
    fn new() -> Runner {
        Runner {
            ctx: Context::new(),
            trace: Rc::new(RefCell::new(Vec::new())),
            fibers: Rc::new(RefCell::new(HashMap::new())),
            plugins: Rc::new(RefCell::new(HashMap::new())),
            children: HashMap::new(),
        }
    }

    /// The cached plugin identity for `name`, built on first use. Caching
    /// pins the registry identity across multiple starts.
    fn plugin(&self, name: &str, params: &Params) -> Rc<FnPlugin<i32>> {
        if let Some(p) = self.plugins.borrow().get(name) {
            return Rc::clone(p);
        }
        let spawns = self.children.get(name).cloned().unwrap_or_default();
        let mut p = make_plugin(name, params.lifo, spawns, &self.trace, &self.fibers);
        let deps: Vec<&str> = params.deps.iter().map(String::as_str).collect();
        if !deps.is_empty() {
            p = p.inject(&deps);
        }
        let p = Rc::new(p);
        self.plugins.borrow_mut().insert(name.to_string(), Rc::clone(&p));
        p
    }

    fn run(&mut self, line: &str) {
        let mut tokens = line.split_whitespace().map(String::from);
        let op = tokens.next().expect("op").clone();
        let args: Vec<String> = tokens.collect();
        let params = parse_params(&args[1.min(args.len())..]);

        match op.as_str() {
            "provide" | "provide-in-realm" => {
                let scope = if op == "provide-in-realm" {
                    self.ctx.isolate_shared(&args[0], &params.realm)
                } else {
                    self.ctx.clone()
                };
                let provider = self.provider_for(&args[0]);
                let fiber = start_fn(&scope, &provider, 0).expect("start provider");
                self.fibers.borrow_mut().insert(format!("provider:{}", args[0]), fiber);
                self.trace.borrow_mut().push(format!("provided {}", args[0]));
            }
            "withdraw" | "withdraw-in-realm" => {
                let fiber = self.fibers.borrow_mut().remove(&format!("provider:{}", args[0])).expect("provider fiber");
                fiber.dispose();
                self.trace.borrow_mut().push(format!("withdrawn {}", args[0]));
            }
            "start" | "start-isolated" => {
                let mut scope = self.ctx.clone();
                if op == "start-isolated" {
                    for dep in &params.deps {
                        scope = scope.isolate_shared(dep, &params.realm);
                    }
                }
                let p = self.plugin(&args[0], &params);
                let fiber = start_fn(&scope, &p, params.config).expect("start");
                self.fibers.borrow_mut().insert(args[0].clone(), fiber);
            }
            "spawn" => {
                let parent = params.realm.clone(); // parent=<name> parsed as realm
                let spawn = Spawn { name: args[0].clone(), deps: params.deps.clone(), config: params.config };
                self.children.entry(parent).or_default().push(spawn);
            }
            "delete" => {
                let p = self.plugin(&args[0], &Params::default());
                self.ctx.registry().delete(&p);
                self.trace.borrow_mut().push(format!("deleted {}", args[0]));
            }
            "expect-registry-size" => {
                let want: usize = args[0].parse().expect("size int");
                let got = self.ctx.registry().size();
                assert_eq!(got, want, "expected registry size {want}, got {got}");
                self.trace.borrow_mut().push(format!("registry-size {want}"));
            }
            "update" => {
                let fiber = self.fibers.borrow().get(&args[0]).cloned().expect("fiber");
                let config: i32 = args[1].parse().expect("config int");
                fiber.update(value(config)).expect("update");
            }
            "restart" => {
                self.fibers.borrow().get(&args[0]).cloned().expect("fiber").restart().expect("restart");
            }
            "dispose" => {
                self.fibers.borrow().get(&args[0]).cloned().expect("fiber").dispose();
                self.trace.borrow_mut().push(format!("disposed {}", args[0]));
            }
            "restart-root" => {
                self.ctx.fiber().dispose();
                self.trace.borrow_mut().push("root-restarted".to_string());
            }
            "expect-state" => {
                let fiber = self.fibers.borrow().get(&args[0]).cloned().expect("fiber");
                let got = state_name(fiber.state());
                assert_eq!(
                    got, args[1],
                    "expected {} {}, trace:\n{}",
                    args[0],
                    args[1],
                    self.trace.borrow().join("\n")
                );
                self.trace.borrow_mut().push(format!("state {} {}", args[0], args[1]));
            }
            other => panic!("unknown op {other}"),
        }
    }

    fn provider_for(&self, service: &str) -> FnPlugin<i32> {
        let service = service.to_string();
        plugin(&format!("provider:{service}"), move |ctx: &Context, _: &i32| {
            ctx.provide_named(&service, value(1i32))?;
            Ok(())
        })
    }
}

fn run_lifecycle_scenario(scenario_file: &str, expected_file: &str) {
    let scenario = read_lines(scenario_file);
    let expected = read_lines(expected_file);

    let mut r = Runner::new();
    for line in &scenario {
        r.run(line);
    }

    let trace = r.trace.borrow();
    compare_trace(&trace, &expected);
}

#[test]
fn golden_scenario() {
    run_lifecycle_scenario("scenario.txt", "expected.txt");
}

#[test]
fn golden_scenario_cascade() {
    run_lifecycle_scenario("scenario-cascade.txt", "expected-cascade.txt");
}

fn event_listener(trace: &Trace, event: &str, who: &str) -> Listener {
    let trace = Rc::clone(trace);
    let prefix = format!("fired {event} {who} payload=");
    Rc::new(move |args: &[Value]| {
        let payload = args[0].downcast_ref::<i32>().copied().unwrap_or(0);
        trace.borrow_mut().push(format!("{prefix}{payload}"));
        None
    })
}

#[test]
fn golden_scenario_events() {
    let scenario = read_lines("scenario-events.txt");
    let expected = read_lines("expected-events.txt");

    let ctx = Context::new();
    let trace: Trace = Rc::new(RefCell::new(Vec::new()));

    for line in &scenario {
        let tokens: Vec<&str> = line.split_whitespace().collect();
        let op = tokens[0];
        let event = tokens[1];
        let mut payload = 0i32;
        let mut realm = String::new();
        for tok in &tokens[2..] {
            if let Some(rest) = tok.strip_prefix("payload=") {
                payload = rest.parse().expect("payload int");
            } else if let Some(rest) = tok.strip_prefix("realm=") {
                realm = rest.to_string();
            }
        }
        match op {
            "on" | "on-isolated" => {
                let (scope, who) = if realm.is_empty() {
                    (ctx.clone(), "root".to_string())
                } else {
                    (ctx.isolate_shared(event, &realm), format!("realm={realm}"))
                };
                scope
                    .on_named(event, event_listener(&trace, event, &who), EventOptions::default())
                    .expect("on");
            }
            "on-global" => {
                ctx.on_named(
                    event,
                    event_listener(&trace, event, "global"),
                    EventOptions { global: true, ..EventOptions::default() },
                )
                .expect("on-global");
            }
            "emit" => ctx.emit_named(event, &[value(payload)]),
            "emit-filtered" => {
                let scope = ctx.isolate_shared(event, &realm);
                let emitter = scope.with_filter(scope.realm_filter(event));
                emitter.emit_named(event, &[value(payload)]);
            }
            other => panic!("unknown op {other}"),
        }
    }

    let got = trace.borrow();
    compare_trace(&got, &expected);
}

#[test]
fn golden_dsl_parse_params() {
    let p = parse_params(&["inject=a,,b".to_string(), "realm=t".to_string(), "config=7".to_string(), "lifo".to_string(), "ignored".to_string()]);
    assert_eq!(p.deps, vec!["a", "b"]);
    assert_eq!(p.realm, "t");
    assert_eq!(p.config, 7);
    assert!(p.lifo);

    let empty = parse_params(&[]);
    assert!(empty.deps.is_empty());
    assert_eq!(empty.config, 0);
    assert!(!empty.lifo);

    // Malformed integers surface as a panic instead of a confusing trace
    // mismatch.
    let result = std::panic::catch_unwind(|| parse_params(&["config=nope".to_string()]));
    assert!(result.is_err(), "malformed config must panic");
}

#[test]
fn golden_dsl_tokenize() {
    let (op, args) = tokenize("  start   worker inject=a config=1 ");
    assert_eq!(op, "start");
    assert_eq!(args, vec!["worker", "inject=a", "config=1"]);
}
